tcpdirect.cu

```diff
@@ -0,0 +1,785 @@
+      sprintf(ethtool_cmd, "ethtool --set-rxfh-indir %s equal 8", opts->tcpdirect_link_name);
+      ret = ret | system(ethtool_cmd);
+      printf("ethtool cmds returned %i, sleeping 1...\n", ret);
+      sleep(1);
+    }
+  }
+
+  *f_mbuf = tmbuf;
+  tmbuf->gpu_mem_fd_ = gpu_mem_fd_;
+  tmbuf->dma_buf_fd_ = dma_buf_fd_;
+  tmbuf->gpu_gen_mem_ = gpu_gen_mem_;
+  tmbuf->cpy_buffer = malloc(opts->buffer_size);
+  tmbuf->vectors = new std::vector<devmemvec>();
+  tmbuf->tokens = new std::vector<devmemtoken>();
+  tmbuf->bytes_received = 0;
+  tmbuf->bytes_sent = 0;
+
+  cudaMalloc(&tmbuf->gpu_rx_mem_, opts->buffer_size);
+  cudaMalloc(&tmbuf->gpu_scatter_list_, opts->buffer_size);
+  tmbuf->rx_blks_ = new std::vector<TcpDirectRxBlock>();
+  tmbuf->scattered_data_ = new std::vector<long3>();
+  return 0;
+}
+
+int udmabuf_setup_alloc(const struct options *opts, void **f_mbuf) {
+  bool is_client = opts->client;
+  int devfd;
+  int memfd;
+  int buf;
+  int buf_pages;
+  int ret;
+  size_t size = opts->tcpdirect_phys_len;
+
+  struct tcpdirect_udma_mbuf *tmbuf;
+  struct dma_buf_create_pages_info pages_create_info;
+  struct udmabuf_create create;
+
+  if (f_mbuf == NULL) return ENOMEM;
+
+  if (*f_mbuf) return 0;
+
+  tmbuf = (struct tcpdirect_udma_mbuf *)calloc(1, sizeof(struct tcpdirect_udma_mbuf));
+  if (!tmbuf) {
+    exit(EXIT_FAILURE);
+  }
+
+  devfd = open("/dev/udmabuf", O_RDWR);
+  if (devfd < 0) {
+    printf("%s: [skip,no-udmabuf: Unable to access DMA buffer device file]\n",
+           TEST_PREFIX);
+    exit(70);
+  }
+
+  memfd = memfd_create("udmabuf-test", MFD_ALLOW_SEALING);
+  if (memfd < 0) {
+    printf("%s: [skip,no-memfd]\n", TEST_PREFIX);
+    exit(72);
+  }
+
+  ret = fcntl(memfd, F_ADD_SEALS, F_SEAL_SHRINK);
+  if (ret < 0) {
+    printf("%s: [skip,fcntl-add-seals]\n", TEST_PREFIX);
+    exit(73);
+  }
+
+  ret = ftruncate(memfd, size);
+  if (ret == -1) {
+    printf("%s: [FAIL,memfd-truncate]\n", TEST_PREFIX);
+    exit(74);
+  }
+
+  memset(&create, 0, sizeof(create));
+
+  create.memfd = memfd;
+  create.offset = 0;
+  create.size = size;
+  printf("size=%lu\n", size);
+  buf = ioctl(devfd, UDMABUF_CREATE, &create);
+  if (buf < 0) {
+    printf("%s: [FAIL, create udmabuf]\n", TEST_PREFIX);
+    exit(75);
+  }
+
+  pages_create_info.dma_buf_fd = buf;
+  pages_create_info.create_page_pool = is_client ? 0 : 1;
+
+  /* TODO: hardcoded NIC pci address */
+  // "0000:06:00.0"
+  ret = sscanf(opts->tcpd_nic_pci_addr, "0000:%llx:%llx.%llx",
+         &pages_create_info.pci_bdf[0],
+         &pages_create_info.pci_bdf[1],
+         &pages_create_info.pci_bdf[2]);
+
+  if (ret != 3) {
+    printf("%s: [FAIL, parse fail]\n", TEST_PREFIX);
+    exit(76);
+  }
+
+  buf_pages = ioctl(buf, DMA_BUF_CREATE_PAGES, &pages_create_info);
+  if (buf_pages < 0) {
+    perror("ioctl DMA_BUF_CREATE_PAGES: [FAIL, create pages fail]\n");
+    exit(77);
+  }
+
+  if (!is_client) {
+    /* TODO hardcoded num_queues */
```

Comment:
Is this active code? The num_queues looks hardcoded, but I thought that was a config passed.